### ISSUE TITLE
[High] Running a Loop on a Job trigger by another Job bugs

### ISSUE THREAD

**Issue by amartinez1660** [RockTomate/Issues#20]
**Steps to Reproduce**
* Create a Job with a loop (I used %CommandLineArgs%)
* Make a for loop with the resulting items collection.

**Expected Results**
To iterate through the items and it does when on the job itself.

**Actual Results**
When the job is triggered stand alone it works, when it's triggered by another process, it fails.

**Comment by ElmarTalibzade** [comment]
Hi,

> Make a for loop with the resulting items collection.

Wanted to ask what do you mean by that?

Would making a Loop that prints `%CommandLineArgs%` be sufficient to reproduce this problem?

> When the job is triggered stand alone it works, when it's triggered by another process, it fails.

If I understand correctly, the Job fails if I run it externally, correct?

Elmar

**Comment by amartinez1660** [comment]
My bad, redaction was horrible, let me clarify again:

> Make a for loop with the resulting items collection.
> Would making a Loop that prints
Yes, printing should be sufficient, although I remember now, I create variables inside the Job too.
* Create a variable, I used "isOSX"
* Add a **Loop** step to the job
* For the iteration `%items%` use any Macro that creates the requried IEnumerable, I tested actually with `split(get(%CommandLineArgs%  0), "/")`
* Print the %item% given by the for loop
* if it finds a string I set the variable to 'true'

Use the 'true' variable somewhere else later.

>If I understand correctly, the Job fails if I run it externally, correct?

Exactly, in the above example, when the job is executed by starting it directly it will work.
When started by another RunJob (not RunJob External) externally it didn't.

Hope that clarifies a bit more.

**Comment by amartinez1660** [comment]
if it is of any help, the error at hand is like this:
`NullReferenceException: Object reference not set to an instance of an object
HardCodeLab.RockTomate.Steps.LoopStep.set_DoLoop (System.Boolean value) (at Assets/RockTomate/Scripts/Steps/Classes/Control Flow/LoopStep.cs:57)
HardCodeLab.RockTomate.Steps.LoopStep.OnReset () (at Assets/RockTomate/Scripts/Steps/Classes/Control Flow/LoopStep.cs:86)
HardCodeLab.RockTomate.Core.Steps.Step.Reset () (at Assets/RockTomate/Scripts/Core/Steps/Step.cs:351)
HardCodeLab.RockTomate.Core.Steps.Scope+<>c.<Reset>b__26_0 (HardCodeLab.RockTomate.Core.Steps.Step x) (at Assets/RockTomate/Scripts/Core/Data/Session/Scope.cs:148)
System.Collections.Generic.List`1[T].ForEach (System.Action`1[T] action) (at <437ba245d8404784b9fbab9b439ac908>:0)
HardCodeLab.RockTomate.Core.Steps.Scope.Reset () (at Assets/RockTomate/Scripts/Core/Data/Session/Scope.cs:148)
HardCodeLab.RockTomate.Steps.GroupStep.OnReset () (at Assets/RockTomate/Scripts/Steps/Classes/General/GroupStep.cs:19)
HardCodeLab.RockTomate.Core.Steps.Step.Reset () (at Assets/RockTomate/Scripts/Core/Steps/Step.cs:351)
HardCodeLab.RockTomate.Core.Steps.Step+<Execute>d__45.MoveNext () (at Assets/RockTomate/Scripts/Core/Steps/Step.cs:294)
HardCodeLab.RockTomate.Core.Coroutines.EditorCoroutines.MoveNext (HardCodeLab.RockTomate.Core.Coroutines.ECoroutine coroutine) (at Assets/RockTomate/Scripts/Core/Utils/EditorCoroutines/EditorCoroutines.cs:202)
HardCodeLab.RockTomate.Core.Coroutines.EditorCoroutines.OnUpdate () (at Assets/RockTomate/Scripts/Core/Utils/EditorCoroutines/EditorCoroutines.cs:185)
System.Reflection.MonoMethod.Invoke (System.Object obj, System.Reflection.BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) (at <437ba245d8404784b9fbab9b439ac908>:0)
Rethrow as TargetInvocationException: Exception has been thrown by the target of an invocation.
System.Reflection.MonoMethod.Invoke (System.Object obj, System.Reflection.BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) (at <437ba245d8404784b9fbab9b439ac908>:0)
System.Reflection.MethodBase.Invoke (System.Object obj, System.Object[] parameters) (at <437ba245d8404784b9fbab9b439ac908>:0)
System.Delegate.DynamicInvokeImpl (System.Object[] args) (at <437ba245d8404784b9fbab9b439ac908>:0)
System.MulticastDelegate.DynamicInvokeImpl (System.Object[] args) (at <437ba245d8404784b9fbab9b439ac908>:0)
System.Delegate.DynamicInvoke (System.Object[] args) (at <437ba245d8404784b9fbab9b439ac908>:0)
UnityEditor.EditorApplication.Internal_CallUpdateFunctions () (at <bf679006b1b84db2a5a44842ef13dc36>:0)`

**Comment by ElmarTalibzade** [comment]
Fixed in 1.0.3

**Comment by BaalEvan** [comment]
I Think this should be Reopened.
Ver 1.1.3 (AssetStore)
Log:
`NullReferenceException: Object reference not set to an instance of an object
HardCodeLab.RockTomate.Steps.LoopStep.set_DoLoop (System.Boolean value) (at Assets/Plugins/RockTomate/Scripts/Steps/Classes/Control Flow/LoopStep.cs:57)
HardCodeLab.RockTomate.Steps.LoopStep.OnReset () (at Assets/Plugins/RockTomate/Scripts/Steps/Classes/Control Flow/LoopStep.cs:86)
HardCodeLab.RockTomate.Core.Steps.Step.Reset () (at Assets/Plugins/RockTomate/Scripts/Core/Steps/Step.cs:362)
HardCodeLab.RockTomate.Core.Steps.Step+<Execute>d__45.MoveNext () (at Assets/Plugins/RockTomate/Scripts/Core/Steps/Step.cs:302)
HardCodeLab.RockTomate.Core.Coroutines.EditorCoroutines.MoveNext (HardCodeLab.RockTomate.Core.Coroutines.ECoroutine coroutine) (at Assets/Plugins/RockTomate/Scripts/Core/Utils/EditorCoroutines/EditorCoroutines.cs:205)
HardCodeLab.RockTomate.Core.Coroutines.EditorCoroutines.OnUpdate () (at Assets/Plugins/RockTomate/Scripts/Core/Utils/EditorCoroutines/EditorCoroutines.cs:188)
System.Reflection.MonoMethod.Invoke (System.Object obj, System.Reflection.BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) (at <fb001e01371b4adca20013e0ac763896>:0)
Rethrow as TargetInvocationException: Exception has been thrown by the target of an invocation.
System.Reflection.MonoMethod.Invoke (System.Object obj, System.Reflection.BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) (at <fb001e01371b4adca20013e0ac763896>:0)
System.Reflection.MethodBase.Invoke (System.Object obj, System.Object[] parameters) (at <fb001e01371b4adca20013e0ac763896>:0)
System.Delegate.DynamicInvokeImpl (System.Object[] args) (at <fb001e01371b4adca20013e0ac763896>:0)
System.MulticastDelegate.DynamicInvokeImpl (System.Object[] args) (at <fb001e01371b4adca20013e0ac763896>:0)
System.Delegate.DynamicInvoke (System.Object[] args) (at <fb001e01371b4adca20013e0ac763896>:0)
UnityEditor.EditorApplication.Internal_CallUpdateFunctions () (at <8004fcc221b54f98ba547350ea71d812>:0)`

Debugging:
RuntimeState is Null for DoLoop, because the step is inside another session (RunJob) and JobSession._stepStates doesn't contain stepStates for childSession steps, also main JobSession doesn't contain a reference to ChildSession if started from RunJobStep.
An additional problem is inside TryGetStepState False Condition if I assign Session from RunJob to MainSession.ChildSession ChildSession isn't Equal to MainSession, and the method return false.

FIX:
I fixed it locally by adding
`context.Session.ChildSession = _childSession;` at RunJobStep:73
then
changing
JobSession:366 from `if (ChildSession == null || !ChildSession.Equals(this))` to `if (ChildSession == null || ChildSession.Equals(this))`
![image](https://user-images.githubusercontent.com/1843494/125700760-39aa0769-e664-4ce1-a79e-e8ef0dbd23cf.png)
![image](https://user-images.githubusercontent.com/1843494/125700778-c87f01a7-7e03-493e-8115-177a40743d5d.png)

**Comment by ElmarTalibzade** [comment]
@BaalEvan Thanks a lot for reporting this and even more thanks for having a fix lined up!

I found a slightly different solution. The solution is converting `DoLoop`, `CurrentIteration` and `TotalItemsCount` properties to normal fields so that we won't be relying on the `RuntimeState` instance. 

Wait, what?

The initial idea was to allow RockTomate to preserve its state at any point in time and be able to resume the Job even when we restart Unity. I eventually decided to shelve this idea for future releases. As a result, there is some unutilized piece of code in a codebase.

I am going to revisit this feature in the future and, hopefully, have better support for preserving states (and the one that will actually be used by RockTomate).

The fix will be included in the next update. I will drop you a DM shortly with a hotfix.
